### ISSUE TITLE
[no-access-state-in-setstate] inspects non-react classes

### ISSUE THREAD

**Issue by fabiospampinato** [yannickcr/eslint-plugin-react#2464]
The `react/no-access-state-in-setstate` rule can find errors in non-react classes, which doesn't make a lot of sense.

In my specific use case I'm using [overstated](https://github.com/fabiospampinato/overstated) for state management, which provides a similar API to React's own API for updating the state, which trips this rule.

It's impossible in general to statically detect if a class is a react component or not, e.g. `class Foo extends Abstract {}`, who knows if that's a react component or not.

Therefor I propose adding an option for blacklisting specific classes, so that I can tell the rule that "Abstract" in my example above is not a react class.

**Comment by ljharb** [comment]
Can you provide some specific code that's detected as a React component?

**Comment by fabiospampinato** [comment]
@ljharb for example:

```js
class Foo extends Abstract {

  update = () => {

    const result = this.getResult ( this.state.foo );

    return this.setState ({ result });

  };

}
```

**Comment by fabiospampinato** [comment]
> Certainly that shouldn't be detected as a React component, thanks.

I mean it depends, how do you know Abstract isn't something that extends `React.Component` for instance?

**Comment by ljharb** [comment]
Sure, but a) it is in *incredibly* discouraged and bad pattern to use inheritance hierarchies in React, so you shouldn't be doing that anyways; and b) there's a jsdoc "extends" comment you're already required to use in those situations to opt your component into being detected by this plugin.
